### PR TITLE
Improve line iterators

### PR DESCRIPTION
**Pull request by rfuest**
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [ ] Check that you've added passing tests and documentation
- [ ] Add a simulator example(s) where applicable
- [ ] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [ ] Run `rustfmt` on the project
- [ ] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This PR started out as a simple refactor of the line iterators to reduce the code duplication introduced when the `Points` iterator was added. But I noticed a problem with the thick line iterators and somehow the scope of the PR widened until it ended up being a complete rewrite of the line iters. The new implementation tries to split the code into more individually testable pieces e.g. by using the `Bresenham` struct to get the start positions of normal and extra parallel lines.

The previous implementation had some problem with the error values on one of the sides, which caused the odd 1px difference in the extra line length. And this also caused some overdraw which could be problematic when we add alpha support. I didn't find the exact cause in the original code, but it should be fixed now and there are a lot of tests for this.

The PR needs some final tweaks before it can be merged (like a changelog entry and maybe some more docs), but I would appreciate some initial feedback. It's easy to miss some corner cases that might be broken now.

Open questions:
* I've changed the code organization by moving each iterator in a separate file in a new `line` directory. By reducing the size of each individual file, especially when there are a lot of tests, this makes the code neater IMO. If you agree I would like to apply the same structure to all primitives (in another PR).
* How exactly should we handle lines with the same start and end point?
  For 1px wide lines a single pixel should be drawn IMO. Otherwise there would be no way of drawing a single pixel with a line primitive, which could lead to problems for e.g. animations.
  But, because the orientation of such a line is unknown, this needs a special case for thick lines. The implementation in this PR assumes that the line is horizontal in this case. Not perfect, but I've seen similar behavior in other APIs. Should we keep this behavior or do you have a better idea?

### CODE REVIEW

File: embedded-graphics/src/geometry/point.rs [hunk 1]
```diff
@@ -176,6 +184,24 @@ impl Point {
...
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// use embedded_graphics::geometry::Point;
+    ///
+    /// let p = Point::new(3, 4);
+    ///
+    /// assert_eq!(p.length_squared(), 25);
+    /// ```
+    pub(crate) fn length_squared(self) -> i32 {
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/geometry/point.rs:202]: It would indeed reduce the number of cast in our code, but I'm not sure if this is true for all code using this method. I would suggest to leave it as `i32`and add a note in #338 to decide this when we make the method public.

File: embedded-graphics/src/primitives/line/thick_points.rs [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                        let error_before_decrease = *error;
+
+                        if self.parallel_parameters.decrease_error(error) {
+                            return (point, error_before_decrease);
+                        }
+                    } else {
+                        if self.parallel_parameters.increase_error(error) {
+                            return (point, *error);
+                        }
+                    };
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/primitives/line/thick_points.rs]: This was intentional to make the cases look the same. But I've changed it anyway.

File: embedded-graphics/src/primitives/line/bresenham.rs [hunk 3]
```diff
@@ -0,0 +1,343 @@
+use crate::{geometry::Point, primitives::Line};
+
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+/// Struct to hold major and minor values.
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/primitives/line/bresenham.rs:4]: `major` isn't necessarily the larger value, but the value used when a major step is taken.

**Comment by rfuest** [comment]
Thanks for the review. The docs really needed some work, but I hope they are better now.

> I think the current behaviour is ok. Can you add a few doc words to `Line` explaining this behaviour? The alternative might be to draw a single dot instead, but that would show as a jump in thickness going from 0 length to 1 length which isn't consistent.

I'm not quite sure where to put this, because the `Line` docs might not be the correct place for this, because this behavior is only relevant for `Styled<Line>`. IMO the (styled) primitives need an overhaul anyway. Can we add this to a new issue?

> ..., with polyline getting a huge 55% boost which is a pleasant surprise.

Nice, that's probably caused by `line::Points` not using `ThickLineIterator` anymore.

> This sums up my thick line impl journey pretty well :grin:

I can understand this very well and don't want to see any line drawing code in the near future either.

File: embedded-graphics/src/primitives/line/thick_points.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{mock_display::MockDisplay, pixelcolor::Gray8};
+
+    /// Draws the output of `ParallelsIterator` to a `MockDisplay`.
+    ///
+    /// Each parallel line is drawn using a different `Gray8` color, to allow testing
+    /// of the drawing order. Points that are drawn multiple times are marked using
+    /// `Gray8::new(0xFF)`.
```
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/primitives/line/thick_points.rs:242]: It might be useful to move this into `mock_display` to make it reusable for other tests.

File: embedded-graphics/src/primitives/line/bresenham.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/primitives/line/bresenham.rs:4]: Did you read the comments I've added to the fields? Are they enough or do you think this needs further explanation?

File: embedded-graphics/src/primitives/line/thick_points.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfuest) [reply embedded-graphics/src/primitives/line/thick_points.rs:242]: Sure, this is outside the scope of this PR.

File: embedded-graphics/src/primitives/line/bresenham.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamwaffles) [reply embedded-graphics/src/primitives/line/bresenham.rs:4]: I did yeah. On second thoughts they're fine as is so I'll mark this comment as resolved to keep the ball rolling.

**Comment by rfuest** [comment]
I'm currently writing one...

**Comment by rfuest** [comment]
Added and ready for another review.

File: embedded-graphics/src/primitives/line/thick_points.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamwaffles) [reply embedded-graphics/src/primitives/line/thick_points.rs:242]: I've opened a ticket in #356 to track this as I've found a need for this functionality and a bit more whilst writing #342
